File: mac/VoiceTests/SyncServiceTests.swift
import XCTest
@testable import Voice

// G5/C6: SyncService is @MainActor.
@MainActor
final class SyncServiceTests: XCTestCase {
    // Tests must not leave UserDefaults modified: snapshot every key `sync()`/`push()` can touch and
    // restore it verbatim in tearDown, regardless of what the test set it to.
    private var savedHotkey: String?
    private var savedMode: String?
    private var savedLanguage: String?

    override func setUpWithError() throws {
        let d = UserDefaults.standard
        savedHotkey = d.string(forKey: Preferences.Key.hotkey)
        savedMode = d.string(forKey: Preferences.Key.mode)
        savedLanguage = d.string(forKey: Preferences.Key.language)
    }

    override func tearDownWithError() throws {
        restore(Preferences.Key.hotkey, savedHotkey)
        restore(Preferences.Key.mode, savedMode)
        restore(Preferences.Key.language, savedLanguage)
    }

    private func restore(_ key: String, _ value: String?) {
        let d = UserDefaults.standard
        if let value { d.set(value, forKey: key) } else { d.removeObject(forKey: key) }
    }

    /// Builds a `MeResponse` by decoding a literal JSON payload shaped like the real wire format
    /// (rather than a memberwise init `MeResponse.Server` no longer has, since it decodes `allowedModels`
    /// tolerantly via a custom `init(from:)`).
    private func meResponse(mode: String, language: String, hotkey: String, llmModel: String? = nil) -> MeResponse {
        let llmModelJSON = llmModel.map { "\"\($0)\"" } ?? "null"
        let json = """
        {
          "user": {"id": "u1", "name": "Miguel"},
          "settings": {"mode": "\(mode)", "language": "\(language)", "hotkey": "\(hotkey)", "llmModel": \(llmModelJSON)},
          "dictionary": [],
          "server": {"version": "0.1.0", "model": "gpt-oss:120b", "concurrency": 1}
        }
        """
        return try! JSONDecoder().decode(MeResponse.self, from: Data(json.utf8))
    }

    // G1: a server-driven hotkey change goes through the callback, never a direct Preferences write.
    func testServerHotkeyChangeGoesThroughCallback() async {
        Preferences.hotkey = .fn
        let api = StubAPI()
        api.meResult = .success(meResponse(mode: "clean", language: "auto", hotkey: "rightOption"))
        let s = SyncService(api: api)
        var changes: [HotkeyChoice] = []
        s.onHotkeyChange = { changes.append($0) }

        await s.sync()

        // HotkeyChoice doesn't declare Equatable (out of this round's touched files) — compare rawValue.
        XCTAssertEqual(changes.map(\.rawValue), ["rightOption"])
        XCTAssertEqual(Preferences.hotkey.rawValue, "fn")   // SyncService itself never wrote it
    }

    // G3(a): push must not clear a server-set llmModel it doesn't know about.
    func testPushPreservesServerLlmModelOnMerge() async {
        let api = StubAPI()
        api.meResult = .success(meResponse(mode: "clean", language: "auto", hotkey: "fn", llmModel: "gpt-oss:120b"))
        let s = SyncService(api: api)
        await s.sync()

        await s.push(mode: "literal")

        XCTAssertEqual(api.puts.count, 1)
        XCTAssertEqual(api.puts.first?.llmModel, "gpt-oss:120b")
        XCTAssertEqual(api.puts.first?.mode, "literal")
    }

    // G3(b): pushing back the value sync() just wrote is a no-op — kills the VoiceApp .onChange echo.
    func testPushIsNoOpWhenUnchanged() async {
        let api = StubAPI()
        api.meResult = .success(meResponse(mode: "clean", language: "auto", hotkey: "fn"))
        let s = SyncService(api: api)
        await s.sync()

        await s.push(mode: "clean")

        XCTAssertEqual(api.puts.count, 0)
    }
}

File: windows/Spit.Core.Tests/SyncServiceTests.cs
using System.Text.Json;

namespace Spit.Core.Tests;

/// Port of mac/VoiceTests/SyncServiceTests.swift. The Mac snapshots and restores UserDefaults around each
/// test; here every test gets its own `InMemoryLocalSettings`, so there is nothing to restore.
public sealed class SyncServiceTests
{
    /// Builds a `MeResponse` by decoding a literal payload shaped like the real wire format, through the
    /// same serializer options the HTTP client uses.
    internal static MeResponse meResponse(string mode, string language, string hotkey, string? llmModel = null)
    {
        var llmModelJson = llmModel is null ? "null" : $"\"{llmModel}\"";
        var json = $$"""
        {
          "user": {"id": "u1", "name": "Miguel"},
          "settings": {"mode": "{{mode}}", "language": "{{language}}", "hotkey": "{{hotkey}}", "llmModel": {{llmModelJson}}},
          "dictionary": [],
          "server": {"version": "0.1.0", "model": "gpt-oss:120b", "concurrency": 1}
        }
        """;
        return JsonSerializer.Deserialize<MeResponse>(json, SpitJson.Options)!;
    }

    // Mac G1: a server-driven hotkey change goes through a callback, never a direct Preferences write.
    // Windows (rule 46) goes one step further — the server's hotkey is the Mac's and is never applied at
    // all, so there is no callback: the local key survives a sync untouched and unwritten.
    [Fact]
    public async Task testServerHotkeyChangeGoesThroughCallback()
    {
        var settings = new InMemoryLocalSettings(hotkey: "rightCtrl");
        var api = new StubApi { MeResult = Result<MeResponse>.Success(meResponse("clean", "auto", "rightOption")) };
        var s = new SyncService(api, settings);

        await s.SyncAsync();

        Assert.Equal("Miguel", s.UserName);           // the sync itself succeeded
        Assert.Equal("rightCtrl", settings.Hotkey);   // SyncService never wrote it
        Assert.Equal(0, settings.HotkeyWrites);
    }

    // G3(a): push must not clear a server-set llmModel it doesn't know about.
    [Fact]
    public async Task testPushPreservesServerLlmModelOnMerge()
    {
        var api = new StubApi { MeResult = Result<MeResponse>.Success(meResponse("clean", "auto", "fn", llmModel: "gpt-oss:120b")) };
        var s = new SyncService(api, new InMemoryLocalSettings());
        await s.SyncAsync();

        await s.PushAsync(mode: "literal");

        var put = Assert.Single(api.Puts);
        Assert.Equal("gpt-oss:120b", put.LlmModel);
        Assert.Equal("literal", put.Mode);
    }

    // G3(b): pushing back the value sync just wrote is a no-op — kills the settings UI's change echo.
    [Fact]
    public async Task testPushIsNoOpWhenUnchanged()
    {
        var api = new StubApi { MeResult = Result<MeResponse>.Success(meResponse("clean", "auto", "fn")) };
        var s = new SyncService(api, new InMemoryLocalSettings());
        await s.SyncAsync();

        await s.PushAsync(mode: "clean");

        Assert.Empty(api.Puts);
    }
}
